Pthreads

juj opened the pull request:
This set of commits implements initial pthreads library support for Emscripten code, and allows native code to utilize multithreading with the pthreads API. This backs on experimental research on the JavaScript SharedArrayBuffer and Atomics APIs from https://docs.google.com/document/d/1NDGA_gZJ7M7w1Bh8S0AoDyEqwDdRh4uSoTPSNn77PFk/edit?usp=sharing , and as such is preliminary and unsupported in other browsers except current Firefox Nightly builds. By default pthreads support is disabled, and the feature must be explicitly enabled by specifying the compiler flag -s USE_PTHREADS=1 and the linker flag -lpthread.

This is for experimental use only at the moment, since none of the threading specification has yet been standardized, and it may change a lot for the final proposal, even to the point of making large parts of this pull request obsolete.

This PR contains some unit tests, but more is available in https://github.com/juj/posixtestsuite (especially see https://github.com/juj/posixtestsuite/commit/ed4bdd6762f90af27754f947d55711ceb8ce3a29 ), which contains ~300 unit tests for POSIX conformance for pthreads.

There is an accompanying pull request to the LLVM backend at https://github.com/kripken/emscripten-fastcomp/pull/67 .

kripken commented on emcc:
```diff
@@ -703,6 +707,7 @@ try:
     pre_js += open(shared.path_from_root('src', 'emrun_prejs.js')).read() + '\n'
     post_js += open(shared.path_from_root('src', 'emrun_postjs.js')).read() + '\n'
 
+
```
newline here seems unnecessary

juj replied:
Ops. Removed.

kripken commented on emscripten.py:
```diff
@@ -1108,6 +1110,8 @@ def make_emulated_param(i):
                                 'shiftLeftByScalar'];
     fundamentals = ['Math', 'Int8Array', 'Int16Array', 'Int32Array', 'Uint8Array', 'Uint16Array', 'Uint32Array', 'Float32Array', 'Float64Array', 'NaN', 'Infinity']
+    if settings['USE_PTHREADS']:
+      fundamentals += ['Atomics', 'SharedInt8Array', 'SharedInt16Array', 'SharedInt32Array', 'SharedUint8Array', 'SharedUint16Array', 'SharedUint32Array', 'SharedFloat32Array', 'SharedFloat64Array']
```
the non-shared are not really needed, I think? if we can remove them, it would avoid wasting minified names in the asm scope (and since they are first, they get the good names like `a,b,c`...)

juj replied:
Good point. I think they are not. I'll remove the non-shared versions.

kripken commented on src/library.js:
```diff
@@ -5256,7 +5597,13 @@ LibraryManager.library = {
     return 0;
   },
-
+  clock_getcpuclockid__deps: ['emscripten_get_now_is_monotonic', '$PROCINFO'],
```
why do we need the monotonic?

juj replied:
Hmm, this is accidental. Removed.

kripken commented on src/library.js:
```diff
@@ -5256,7 +5597,13 @@ LibraryManager.library = {
+  clock_getcpuclockid__deps: ['emscripten_get_now_is_monotonic', '$PROCINFO'],
+  clock_getcpuclockid: function(pid, clk_id) {
+    if (pid < 0) return ERRNO_CODES.ESRCH;
+    if (pid != 0 && pid != PROCINFO.pid) return ERRNO_CODES.ENOSYS;
```
best to do `!==`

juj replied:
Check.

kripken commented on src/parseTools.js:
```diff
@@ -53,7 +53,10 @@ function preprocess(text) {
         } else if (line[2] == 'n') { // include
-          var included = read(line.substr(line.indexOf(' ')+1));
+          var filename = line.substr(line.indexOf(' ')+1);
+          if (filename.indexOf('"') == 0)
```
curly braces

kripken replied:
`===`

kripken commented on src/preamble.js:
```diff
@@ -1427,10 +1445,11 @@ var __ATMAIN__    = []; // functions called when main() is to be run
 var __ATPOSTRUN__ = []; // functions called after the runtime has exited
 
-var runtimeInitialized = false;
+var runtimeInitialized = ENVIRONMENT_IS_PTHREAD; // The runtime is hosted in the main thread, and bits shared to pthreads via SharedArrayBuffer. No need to init again in pthread.
```
got it, but how about setting to false, then if it's a pthread, set it to true a few lines down? that seems clearer, as the semantic meaning of `runtimeInitialized` is not `is a pthread`

juj replied:
Ok.

kripken commented on src/pthread-main.js:
```diff
@@ -0,0 +1,96 @@
+// Pthread Web Worker startup routine:
```
it seems confusing for `pthread-main.js` to be the main code run first by the pthreads. The name could also be understood as the main of all the pthreads, I think.

juj replied:
Hmm, though the comments in the beginning of this file should make it clearer? Or do you have a rename suggestion that you'd prefer?

kripken replied:
I don't have a better idea I'm afraid. I guess leave it unless you do.

kripken commented on src/pthread-main.js:
```diff
@@ -0,0 +1,96 @@
+};
+
+this.onmessage = function(e) {
+  if (e.data.cmd == 'load') { // Preload command that is called once per worker to parse and load the Emscripten code.
```
`===`

juj replied:
Check.

kripken commented on src/pthread-main.js:
```diff
@@ -0,0 +1,96 @@
+// with the main execution thread.
+var buffer;
+
+var threadBlock = 0; // Info area for this thread in Emscripten HEAP (shared). If zero, this worker is not currently hosting an executing pthread.
```
is this related or similar to to thread local storage?

The term "block" seems possibly confusing here, as this isn't something blocking on something else, but a block of info. How about "header" or "info" or "headerPtr" etc.?

juj replied:
This points to the malloced `pthread_t` structure, of which there exists one for each thread. This pointer is also used to denote the ID of the thread. I liked using the term block since from JS perspective it's a pointer to a block of C memory structure, and that term is not a commonly used term elsewhere in the code, so it seemed to give a unique meaning. I can rename it though, perhaps `threadStruct`? or `threadMemoryBlock`? or `threadInfoStruct`?

kripken replied:
`threadInfoStruct` sounds good.

juj replied:
Ok, done.

kripken commented on src/pthread-main.js:
```diff
@@ -0,0 +1,96 @@
+    try {
+      // HACK: Some code in the wild has instead signatures of form 'void *ThreadMain()', which seems to be ok in native code.
+      // To emulate supporting both in test suites, use the following form. This is brittle!
+      if (typeof asm['dynCall_ii'] !== 'undefined')
```
curly braces

kripken commented on src/shell.js:
```diff
@@ -36,7 +36,12 @@ for (var key in Module) {
+// 2) We could be the application main() thread proxied to worker. (with Emscripten -s PROXY_TO_WORKER=1) (ENVIRONMENT_IS_WORKER == true, ENVIRONMENT_IS_PTHREAD == false)
+// 3) We could be an application pthread running in a worker. (ENVIRONMENT_IS_WORKER == true and ENVIRONMENT_IS_PTHREAD == true)
 var ENVIRONMENT_IS_WORKER = typeof importScripts === 'function';
+if (typeof ENVIRONMENT_IS_PTHREAD === 'undefined') ENVIRONMENT_IS_PTHREAD = false; // ENVIRONMENT_IS_PTHREAD=true will have been preset in pthread-main.js.
```
this is going to set a global variable in the global js scope. if we add `var` here that would be avoided. i don't think there would be a problem to do that here?

juj replied:
Yeah, that seems right.

kripken commented on system/include/emscripten/threading.h:
```diff
@@ -0,0 +1,219 @@
+
+typedef union em_variant_val
+{
+	int i;
```
indentation should be 2

juj replied:
Check.

kripken commented on system/include/emscripten/threading.h:
```diff
@@ -0,0 +1,219 @@
+{
+	int function;
+	int operationDone;
+	em_variant_val args[8];
```
maybe have a constant for the 8? that could be checked elsewhere that we don't exceed it?

juj replied:
Check.

kripken commented on system/include/libc/bits/alltypes.h:
```diff
@@ -88,7 +88,11 @@ typedef struct { union { int __i[9]; unsigned __s[9]; } __u; } pthread_attr_t;
 
 #if defined(__NEED_pthread_mutex_t) && !defined(__DEFINED_pthread_mutex_t)
+#ifdef __EMSCRIPTEN__
+typedef struct { union { int __i[7]; void *__p[7]; } __u; } pthread_mutex_t;
```
why is this needed?

juj replied:
The musl algorithm for implementing a mutex differs in semantics for their `_m_lock` field, so I was not able to reuse that field while using the futex API that we have in the pthreads spec. Updated this with a comment. For a while I attempted to adapt our mutex implementation to reuse musl syscalls architecture, but that got hairy with the semaphores implementation that I concluded it's safer to use a separate field for the mutex lock control structure instead.

kripken commented on system/lib/libc/musl/src/thread/__timedwait.c:
```diff
@@ -0,0 +1,82 @@
+	if (1 || pthread_self()->cancelasync == PTHREAD_CANCEL_ASYNCHRONOUS) {
+		do {
+			if (_pthread_isduecanceled(pthread_self())) {
+				// Emscripten-specific return value: The wait was canceled by user calling
```
why is this different from native?

juj replied:
Native implementations are able to pre-empt asynchronous cancellation with signals-like mechanism, but we don't have that ability in JS, so we must poll for such cancellation points ourselves. That requires that we propagate a custom exit value out from this function to distinguish exiting due to being timed out, or being canceled, so that the caller of this wait is able to do proper cleanup and return the correct return value.

kripken commented on system/lib/libc/musl/src/thread/pthread_attr_get.c:
```diff
@@ -0,0 +1,102 @@
+
+int pthread_attr_getstack(const pthread_attr_t *restrict a, void **restrict addr, size_t *restrict size)
+{
+/// XXX musl is not standard-conforming? It should not report EINVAL if _a_stackaddr is zero, and it should
```
is this something you added, or in musl upstream?

juj replied:
Musl implementation was not passing the open posix test suite here, so I changed this implementation so that it does. I asked Musl upstream about this, but the discussion somehow became about "we will have a proper bug tracker set up soon, so you'll be able to report these pedantic cases there".

kripken commented on system/lib/pthread/library_pthread.c:
```diff
@@ -0,0 +1,543 @@
+	return nsecs;
+}
+
+#if 0
```
why is this in an `if 0` and not just removed?

juj replied:
I've gone back and forth a few times between this implementation for mutexes and the musl one, and I find value in this code as a potential alternative to the musl code, so I thought to keep it at hand for reference for now at least.

kripken commented on emcc:
```diff
@@ -750,6 +755,16 @@ try:
+      settings_changes.append('PTHREAD_POOL_SIZE=0')
+    js_libraries.append(shared.path_from_root('src', 'library_pthread.js'))
+  else:
+    js_libraries.append(shared.path_from_root('src', 'library_pthread_stub.js'))
```
let's assert that not `USE_PTHREADS`, which as the docs say, is wrong (`-lpthread` must be used).

juj replied:
I'm not sure I follow this comment.

The usage will be like this: If one wants to use pthreads, one must use the compiler flag `-s USE_PTHREADS=1`, and the linker flag `-lpthread`. Only using one or the other is not enough, since the changes extend to bitcode level during compilation, and at link time as well.

If these flags are not specified, then the old versions of `pthread_xxx` are used, which are the minimal implementation/stubs. Some codebases might not be interested in actual pthreads support, but might still depend on the stub versions to compile. Also our test suite depends on this.

I will have to write up documentation for developers on how to target pthreads, but that will be a bit later.

kripken replied:
How is bitcode generation affected?

juj replied:
Hmm actually, it might not. I'll have to do tests on that. For now I documented that compilation stage should do `-s USE_PTHREADS=1`, and linking stage should do `-lpthread`.

kripken replied:
We have in general tried to make it easy for people to just use the same flags during compilation and during linking, I'd prefer to not deviate from that here. Perhaps just making either of those flags be enough is fine (i.e. they both do the same thing).

juj replied:
Requiring a `-lpthread` parameter during compilation phase would be a complete no-go, so I suppose removing that altogether and just always using `-s USE_PTHREADS=1` would then be the simplest?

kripken replied:
I am also ok if `-lpthread` just does `-s USE_PTHREADS=1` under the hood, then they are basically synonyms. Then people could use the latter everywhere, as in your last comment, but it would also be ok to use `-lpthread` during link. I'm ok either way.

juj replied:
I've changed the code to only use `-s USE_PTHREADS=1` as the only flag for both compile and like times.

kripken commented on emcc:
```diff
@@ -1407,6 +1422,9 @@ try:
       src = None
 
+  if pthreads:
+    shutil.copyfile(shared.path_from_root('src', 'pthread-main.js'), os.path.join(os.path.dirname(os.path.abspath(target)), 'pthread-main.js'))
```
we should document this adding of the file in the `site/` dir somewhere. it's another file people will need to copy around etc.

juj replied:
Good point, that needs to be mentioned.

juj replied:
Added note.

kripken commented on src/library.js:
```diff
@@ -49,6 +49,9 @@ LibraryManager.library = {
   opendir__deps: ['$FS', '__setErrNo', '$ERRNO_CODES', 'open'],
   opendir: function(dirname) {
+#if USE_PTHREADS
+    if (ENVIRONMENT_IS_PTHREAD) return _emscripten_sync_run_in_main_thread_1({{{ cDefine('EM_DEFERRED_OPENDIR') }}}, dirname);
```
these deferred calls are proxied to the main thread? the name seems a little confusing. it's not that we defer them and forget about them, which is what "defer" says to me at least. Also, we use the term "proxying" elsewhere, so `_PROXIED` seems better?

juj replied:
Ok, will rename to `EM_PROXIED_OPENDIR`, `EM_PROXIED_CLOSEDIR` and so on.

kripken commented on src/library_pthread.js:
```diff
@@ -0,0 +1,655 @@
+      // TODO HACK! Replace the _js variant with just _pthread_testcancel:
+      //_pthread_testcancel();
+      __pthread_testcancel_js();
+      // In Main runtime thread, assist pthreads in performing operations that they need to access the Emscripten main runtime for.
```
what is the "main runtime thread"?

juj replied:
main runtime thread is the JS VM thread that hosts the Emscripten runtime (the thread that initialized the heap and the C runtime). So in default build mode, this is the main browser thread, and in -s PROXY_TO_WORKER=1, it is a web worker thread.

kripken replied:
Got it. We should document this model somewhere - wiki page perhaps? or the threads js library file?

juj replied:
I think the user is not affected here until PROXY_TO_WORKER builds with pthreads are supported. I set the compiler to abort now if these are attempted together, and when this feature is implemented, it'll have documentation.

kripken commented on tools/system_libs.py:
```diff
@@ -652,6 +646,17 @@ def create_gl():
+  def create_dlmalloc_singlethreaded():
+    return create_dlmalloc('dlmalloc.bc', ['-O2'])
+
+  def create_dlmalloc_multithreaded():
```
nice solution, we should do the other stuff that hit this issue similarly. e.g. tracing. cc @nickdesaulniers

kripken commented on tests/test_browser.py:
```diff
@@ -2582,3 +2582,93 @@ def test_modularize(self):
         ''' % code)
         self.run_browser('a.html', '...', '/report_result?0')
 
+  # Test that the emscripten_ atomics api functions work.
```
do these tests pass with a warning if there is no pthreads support in the browser?

juj replied:
No, currently these tests will throw an exception and fail. Probably will have to do something about this. It's slightly annoying that the early-out skip exit will have to live e.g. in src/preamble.js since otherwise the code will throw there when it tries to access nonexisting SharedArrayViews.

kripken replied:
Hmm, you could avoid adding anything to `src/` by having an `onerror` handler in the tests, and if the handler sees there are no SharedArrayViews, it reports a result (of pass or warn).

juj replied:
Made the test harness report skip on the pthread tests if the browser doesn't support them. Also, if a user runs a pthread-compiled page in a browser that doesn't support pthreads, the execution will abort immediately at startup.

kripken commented on src/library_pthread.js:
```diff
@@ -0,0 +1,655 @@
+
+    busySpinWait: function(msecs) {
+      var t = performance.now() + msecs;
+      while(performance.now() < t)
```
how about `{}` or `;` on the same line?

kripken commented:
Very impressive work!

Bunch of questions above. I am still getting into the code, the answers will help me.

gunyarakun commented:
I'm interested in running threads with pthread API in emscripten. 

So I tried to build and run a sample pthread program and run posixtestsuite, but I failed to execute.

Could you point what's wrong I did?

I was using the browser FirefoxNightly 39.0a1 (2015-03-23) and emscripten/fastcomp on pthreads branch.

## testing with simple pthread program

I fetched a pthread sample source code from http://en.wikipedia.org/wiki/POSIX_Threads

I built html/js with emcc and run it with emrun.

```
emcc wikipedia.c -lpthread -o wikipedia.html -s TOTAL_MEMORY=268435456 -s PTHREAD_POOL_SIZE=40 -s USE_PTHREADS=1 --emrun 
emrun --browser firefox_nightly wikipedia.html
```

The error below shows up continuously.

```
still waiting on run dependencies:
dependency: pthreads
(end of list)
```

Here is the source code.

```
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
 
#define NUM_THREADS     5
 
void *perform_work(void *argument)
{
   int passed_in_value;
 
   passed_in_value = *((int *) argument);
   printf("Hello World! It's me, thread with argument %d!\n", passed_in_value);
 
   /* optionally: insert more useful stuff here */
 
   return NULL;
}
 
int main(void)
{
   pthread_t threads[NUM_THREADS];
   int thread_args[NUM_THREADS];
   int result_code, index;
 
   // create all threads one by one
   for (index = 0; index < NUM_THREADS; ++index) {
      thread_args[index] = index;
      printf("In main: creating thread %d\n", index);
      result_code = pthread_create(&threads[index], NULL, perform_work, (void *) &thread_args[index]);
      assert(0 == result_code);
   }
 
   // wait for each thread to complete
   for (index = 0; index < NUM_THREADS; ++index) {
      // block until thread 'index' completes
      result_code = pthread_join(threads[index], NULL);
      printf("In main: thread %d has completed\n", index);
      assert(0 == result_code);
   }
 
   printf("In main: All threads completed successfully\n");
   exit(EXIT_SUCCESS);
}
```

## testing with posixtestsuite

I challenged to run tests in posixtestsuite (https://github.com/juj/posixtestsuite)

- Modify emcc/run_emcc_tests.py emrun option '--browser' to use firefox_nightly

```
cmd = ['emrun', '--kill_start', '--kill_exit', '--timeout=15', '--silence_timeout=15', '--browser=firefox_nightly', out]
```

- Install emcc/run_all_tests.py and emcc/run_emcc_tests.py into a directory in $PATH (/usr/local/bin)

- Execute tests in conformance directory.

```
cd conformance/interfaces
run_all_tests.py
```

The test started well and launched Firefox, but the results are fail like following.
```
=== FINISHED ===
3 tests failed:
1-1.c, 2-1.c, 3-1.c
```

juj commented:
@gunyarakun: There is currently an undocumented issue that you will need to work around: in Firefox there is a limit to how many web workers you can spawn. This is bug https://bugzilla.mozilla.org/show_bug.cgi?id=1052398 . To work around this, navigate to `about:config` and find the pref `dom.workers.maxPerDomain` and set it to a big enough value, e.g. `50`. The value should be larger than what you pass in to `-s PTHREAD_POOL_SIZE=x`, or otherwise you will see that the run dependencies will never finish. Setting the value to at least 50 should allow running the open posix test suite as well, since some of those tests use quite large number of simultaneous threads.

juj commented:
@kripken: Thanks for the reviews, I think I should have addressed all of those in the new batch of commits.

gunyarakun commented:
@juj: thank you for your advice. I changed dom.workers.maxPerDomain and toolkit.startup.max_resumed_crashes (for not showing the dialog to be safe mode) and posixtestsuite seemed working well.

But the sample program following doesn't work well. Is my program is wrong or is it a known problem?

```
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>

static void *thread_func(void *vptr_args) {
    for (int i = 0; i < 20; i++) {
        puts("  b");
        sleep(1);
    }

    return NULL;
}

int main(void) {
    pthread_t thread;

    if (pthread_create(&thread, NULL, thread_func, NULL) != 0) {
        return EXIT_FAILURE;
    }

    for (int i = 0; i < 20; i++) {
        puts("a");
        sleep(1);
    }

    if (pthread_join(thread, NULL) != 0) {
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
```

juj commented:
Thanks for the good test case @gunyarakun! I tested that locally and found several small things. Pushed fixes to the pull request now, the last three commmits specifically. With those fixes, the test works for me. Note that unfortunately the prints will not appear in the text box on the page until the code exists main(), since the DOM changes will only take effect once the main thread yields back to the browser. If you open the page JS console, you can see the prints come in in real time though.

Also, this pull request was now rebased to on top of the latest incoming with the LLVM 3.6 merge that occurred yesterday.

gunyarakun commented:
@juj: Thanks for your quick response and commits. I tried to run my test program with the pthreads branch based on emscripten 1.30.1. It works well :smile:

httpdigest commented:
I would like to use pthread with emscripten as well, but am a bit confused about the required compiler as well as linker flags. GCC used to have the "-pthread" compiler as well as linker flag, which both enables pthread preprocessor/include statements as well as enable linking against pthreads (that's what -lpthread is doing).
So, what should I do when building a simple pthread threaded application with emscripten?
Use -lpthread only during linking and nothing during compiling, or use the GCC standard -pthread for both compiling and linking?

juj commented:
The documentation for this branch lives in https://github.com/juj/emscripten/blob/pthreads/site/source/docs/porting/pthreads.rst , so please follow the build flags instructed there. If the flags change, I'll update the documentation accordingly.

httpdigest commented:
Wow, thanks for the quick answer and the documentation. I would of course be happy to follow the suggested way with the proposed flags. Just wanted to see whether existing config and Makefile scripts can be left unchanged when switching from GCC compilation to emscripten.

httpdigest commented:
Using the example pthread program posted by @gunyarakun I am now getting an assertion failed error at the assert in Runtime.staticAlloc
called from:
allocate <- PThread.allocateUnusedWorkers <- PThread.getNewWorker
It seems that the variable 'staticSealed' is true whereas the assertion expects it to be false.

juj commented:
@httpdigest : Ops sorry, forgot to push the latest few commits that fixed that issue (I'm working on this pull request live). See https://github.com/juj/emscripten/commit/ad704ef55a3a4890a3c71f684889e3698f279674

httpdigest commented:
Cool, thanks! Now this error is gone, but still no thread is being created. The join now waits forever and blocks the rest of the site and Nightly then complaining about the site taking too long.
I reduced the thread_func of above example to a simple puts("Hello") but that does not show.
Only the console is printing:
"Preallocating 1 workers for a pthread spawn pool."
but no thread is being started/executed.
Can you confirm that the above test does work for you?

juj commented:
If you are referring to this test https://github.com/kripken/emscripten/pull/3266#issuecomment-85225571 , since the test creates 5 threads and then _synchronously_ proceeds to wait for them to join (without yielding back to the browser regularly in between), you must compile it with `-s PTHREAD_POOL_SIZE=5` command line flag. This ensures that 5 workers are precreated before main() is called. Tested this to work with `./em++ a.cpp -o a.html -O3 -lpthread -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=5`.

If you are referring to this test https://github.com/kripken/emscripten/pull/3266#issuecomment-85610859 ,  then you must wait for 20 seconds like commented in https://github.com/kripken/emscripten/pull/3266#issuecomment-86002398 , since the prints will only be observable after returning from main(). Tested this to work with `./em++ a.cpp -o a.html -O3 -lpthread -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=1`. Try removing the `sleep()`s in that test to make it finish immediately.

httpdigest commented:
Thanks for the info with the printf not being visible after main returns. I was using the test posted by @gunyarakun here in this comments above.

Currently, I have it stripped down to this:

```C
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

static void *thread_func(void *vptr_args) {
    puts("RUN!");
    return NULL;
}

int main(void) {
    pthread_t thread;
    if (pthread_create(&thread, NULL, thread_func, NULL) != 0) {
        puts("pthread_create did not work!");
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}
```

And "RUN!" is never being printed. Neither is "pthread_create did not work!". Only
"Preallocating 1 workers for a pthread spawn pool." is printed.

juj commented:
Ahh.. Yeah, I can verify. That's actually an annoying case that occurs mostly with toy scenarios only.

What happens is that when main() exits, it immediately tears down the C runtime, and due to scheduling, the thread has not had a chance to run, so when it calls `puts()`, all the file descriptors are already closed and the print is ignored.

To keep the runtime alive, you have two options:
   - Call `emscripten_set_main_loop` to set up a periodic event handler, which signals Emscripten to understand that the runtime should not be torn down when main exits, or perhaps just simpler
   - pass the linker flag `-s NO_EXIT_RUNTIME=1` to command line when building. That forces Emscripten to keep the C runtime alive after exiting from `main()`. Adding that flag, I see the print come out from the thread correctly.

I'll add a documentation to pthreads for this scenario and the command line flag.

httpdigest commented:
Nice! That trick with NO_EXIT_RUNTIME=1 did solve the problem.
Now, is it expected that pthread_join() hangs indefinitely (see my previous edited comment)?

juj commented:
Yes, if you don't use `-s PTHREAD_POOL_SIZE=1`, but synchronously have a sequence

   pthread_create(...)
   pthread_join(...)

without regularly yielding back to browser between create and join, the join will hang. This is because without pooling workers up front, the thread will not have a chance to run immediately on `pthread_create`, but it will have to wait until main yields back to browser. The root cause of this is https://bugzilla.mozilla.org/show_bug.cgi?id=1049079 . Therefore it's a good idea for now to always pool up enough threads with that linker flag to make such code work correctly. It is uncertain if bug 1049079 will ever be fixed, but the need of pooling will go away with another proxying architecture we are currently developing that is not yet in place. In the proxying arch, the whole code will run in a worker, and the main thread would always be responsive to process worker startups, so no pooling would be needed.

httpdigest commented:
Many thanks for this explanation and your help so far!
This is all very exciting and I am looking forward to the point where any (threaded) language can be compiled and run on JavaScript/asm.js.
I stay tuned for your proposed changes.

httpdigest commented:
Hm... I have just switched from puts to printf and now nothing is printed...

```C
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

static void *thread_func(void* vptr_args) {
    printf("RUN!");
  // or printf("%s", "RUN!");
    return NULL;
}

// main is like above
```

Is printf generally not supported in pthreaded code? Somehow only puts works, but puts does not support formatting with variables.

juj commented:
That issue is not related to pthreads, but the stdout and stderr in Emscripten are flushed on a line basis, so you need to add `\n` at the end of the printf to make it show up. `puts` is different that it includes its own newlines.

httpdigest commented:
Ah, I did a fflush(stdout) but that does not seem to have an effect.

juj commented:
Yeah, we had a pull request that would add `fflush`, see #2771, but that caused a lot of headache.

httpdigest commented:
Do you know why the following code does not work? printf does not work for me at all, currently, unless it is automatically optimized to puts, it seems:

```C
#include <stdio.h>
#include <pthread.h>

static void *thread_func(void* vptr_args) {
    printf("%s\n", "Hello");
    return NULL;
}

int main(int argc, char** argv) {
    pthread_t thread;
    pthread_create(&thread, NULL, thread_func, NULL);
    return 0;
}
```

It prints the following:
Preallocating 1 workers for a pthread spawn pool.
(a funny rectangle character with [0, 0, 1, 0] in it)

I only do this toy samples, because I want to make sure, that all low level bits are working before porting JamVM. :)

juj commented:
Thanks! I was able to reproduce, and pushed in a fix. Portions of the main C runtime are not currently multithreaded at all, so we need to proxy operations from pthreads to the main thread. However I was not handling varargs in printf functions correctly. Fixed in https://github.com/juj/emscripten/commit/fe37564a1d1adca2fa374176697ca2351a59b653 now.

Tested your code to work with `./em++ pt.cpp -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=1 -s NO_EXIT_RUNTIME=1 -lpthread -o pt.html`. Toy samples are good, they show faults in controlled conditions. :)

Note that after updating this branch, you will need to clear cache with `./em++ --clear-cache` so that the above commit is rebuilt to take effect.

httpdigest commented:
Nice, speedy fix, thanks! Confirm it works now.

httpdigest commented:
I have yet another question about SharedArrayBuffers and XHR requests.
If I were to do a synchronous XHR request in a Web Worker / pthread with responseType "arraybuffer", this gives me a normal ArrayBuffer. If I now want to allocate a memory region in the SharedArrayBuffer of the C runtime and transfer the received ArrayBuffer into that SharedArrayBuffer region, how would I be doing that? Is there a bulk copy operation or do I have to use a for loop and transfer each byte individually from one to the other buffer?

httpdigest commented:
Is it forseeable when your pthreads branch might make its way into kripken:incoming?
Maybe with a compile-time switch of `-lpthread` so as to switch between SharedArrayBuffer and normal ArrayBuffer code?

juj commented:
Yeah, the intent of this pull request is to get it in. I still have a few fixes to do based on the reviews here, and I'm currently in the process of proofing this in a large real world codebase. The way the code is laid out that if the user does not specify `-s USE_PTHREADS=1`, then Emscripten will generate code that doesn't reference SharedArrayBuffer, Atomics or even `ENVIRONMENT_IS_PTHREAD` anywhere so that users who don't use this feature won't be impacted the slightest.

fchiba commented:
Hi, I found a problem about spawning threads from sub-threads.

  #include <stdio.h>
  #include <pthread.h>

  static void *thread2_func(void *vptr_args) {
    puts("c");
    return NULL;
  }

  static void *thread_func(void *vptr_args) {
    pthread_t thread;
    puts("b");
    pthread_create(&thread, NULL, thread2_func, NULL);
    pthread_join(thread, NULL);
    return NULL;
  }

  int main(void) {
    pthread_t thread;
    puts("a");
    pthread_create(&thread, NULL, thread_func, NULL);
    pthread_join(thread, NULL);

    return 0;
  }

The results should be "abc", but actually it was "ab" and freeze (thread2_func is never called). It seems that postMessage called from pthread_create is blocked by pthread_join.

juj commented:
Thanks @fchiba! Nested pthread spawning requires special support, and that support did not account for the case that the threads might be synchronously joined on. Fixed this in https://github.com/juj/emscripten/commit/e53e7cbccfa5f985917add258e9aecc60ce8b035 and added your snippet as a test case. Note that if you update the branch, you'll need to call `emcc --clear-cache` before building. Also I have rebased both this and the emscripten-fastcomp on top of a more recent upstream.

juj commented:
Oh sorry @httpdigest, I had not noticed your question about the typed arrays in workers:

    I have yet another question about SharedArrayBuffers and XHR requests.
    If I were to do a synchronous XHR request in a Web Worker / pthread
    with responseType "array buffer", this gives me a normal ArrayBuffer. If I
    now want to allocate a memory region in the SharedArrayBuffer of the
    C runtime and transfer the received ArrayBuffer into that SharedArrayBuffer
    region, how would I be doing that? Is there a bulk copy operation or do I
    have to use a for loop and transfer each byte individually from one to
    the other buffer? The SharedArrayBuffer spec does not seem to support
    such a bulk copy operation.

It is possible to call `.set()` on a typed array view to transfer bytes between a shared and a non-shared typed array. See this recent commit that does something like that: https://github.com/juj/emscripten/commit/2b8aea3cd254e005314b3f5033cee2c313133ba1#diff-5df1ca8b00952a4230cf0be8d8bf3682R1156

kripken commented on emcc:
```diff
@@ -735,6 +735,20 @@ try:
 
+  pthreads = 'USE_PTHREADS=1' in settings_changes
+  if pthreads:
+    settings_changes.append('USE_PTHREADS=1')
```
this line seems unnecessary, since we reach it only if that was true anyhow?

in general, I think this code should be moved to around line 996, **after** the `settings_changes` have been applied. Then just test `Settings.USE_PTHREADS`.

juj replied:
Thanks, done.

kripken commented on src/library_pthread.js:
```diff
@@ -0,0 +1,683 @@
+
+      // Allocate memory for thread-local storage.
+      var tlsMemory = allocate({{{ cDefine('PTHREAD_KEYS_MAX') }}} * 4, "i32*", ALLOC_STATIC);
+      for(var i = 0; i < {{{ cDefine('PTHREAD_KEYS_MAX') }}}; ++i) HEAPU32[tlsMemory/4+i] = 0;
```
space after `for`

juj replied:
Done.

kripken commented on src/preamble.js:
```diff
@@ -1127,8 +1135,58 @@ if (totalMemory !== TOTAL_MEMORY) {
+HEAPU32 = new SharedUint32Array(buffer);
+HEAPF32 = new SharedFloat32Array(buffer);
+HEAPF64 = new SharedFloat64Array(buffer);
+#else
```
please add a comment `// USE_PTHREADS`, this is a big `if-else`

kripken commented on src/preamble.js:
```diff
@@ -1501,5 +1594,48 @@ function lookupSymbol(ptr) { // for a pointer, print out all symbols that resolv
 var memoryInitializer = null;
 
-// === Body ===
+#if USE_PTHREADS
```
how about if we have a file `src/pthreads_support.js` which we paste in, in `jsifier.js`, if pthreads are used? Just to avoid this and maybe some other big ifdefs that are pthreads-only? i don't feel strongly on this, just wondering if it would be cleaner that way.

juj replied:
I don't expect that the piece of code in this location would grow much, since it's just to avoid the lack of support for `navigator.hardwareConcurrency` in the browser.

kripken commented on src/shell.js:
```diff
@@ -36,7 +36,18 @@ for (var key in Module) {
 // *** Environment setup code ***
 var ENVIRONMENT_IS_WEB = typeof window === 'object';
 var ENVIRONMENT_IS_NODE = typeof process === 'object' && typeof require === 'function' && !ENVIRONMENT_IS_WEB;
+// Three configurations we can be running in:
```
nice summary

kripken commented on tests/test_browser.py:
```diff
@@ -2512,5 +2512,126 @@ def test_dynamic_link_glemu(self):
     ''')
     Popen([PYTHON, EMCC, 'side.cpp', '-s', 'SIDE_MODULE=1', '-O2', '-o', 'side.js']).communicate()
 
-    self.btest(self.in_dir('main.cpp'), '1', args=['-s', 'MAIN_MODULE=1', '-O2', '-s', 'LEGACY_GL_EMULATION=1', '--pre-js', 'pre.js'])
```
this test appears to have been collateral damage of a merge...

juj replied:
Oops, very sorry!

kripken commented on src/library.js:
```diff
@@ -6264,16 +6536,25 @@ LibraryManager.library = {
   // netdb.h
   // ==========================================================================
 
+#if USE_PTHREADS
```
this pattern seems to show up a lot. it seems there is unnecessary duplication in 2 areas: first, the allocate (or whatever the actual creation is) appears in two places, in both sides of the ifdef. second, the check if is pthread, then use it from PthreadWorkerInit, etc.

How about ` __h_errno_state: {{{ pthreadify('allocate(1, "i32", ALLOC_STATIC)') }}}`, where `pthreadify` does all the stuff automatically?

juj replied:
All these lines in src/library.js are more of a temporary thing that I want to work on removing in favor of a properly multithreaded C library. There is a noticeable perf impact, and a chance of deadlocking even(!) with this proxying scheme, so I'm not sure if this would warrant any formalization, but it's best to just get rid of it. Also it would be a bit complex, since there would be nested calls to {{{ }}}, so would have to implement nesting level detection, e.g.

    tzname: '; if (ENVIRONMENT_IS_PTHREAD) _tzname = PthreadWorkerInit._tzname; else PthreadWorkerInit._tzname = _tzname = allocate({{{ 2*Runtime.QUANTUM_SIZE }}}, "i32*", ALLOC_STATIC)',

kripken replied:
fair enough

kripken commented:
Some general notes:

* libc proxying - will need to be redone if we do syscalls. not sure if we will, there is an experiment now. but i suppose if we do syscalls eventually, we would need to proxy them as we are proxying libc calls, is that right?
* pthreads and shared modules - pretty sure it won't work. `src/*_sharedlib.js` may need updates. can leave as a todo, but need an error in emcc now
* there is no way to test pthreads in the shell, i guess?
* just to be sure: musl bits are 1.0.3 release, no changes unless inside `__EMSCRIPTEN`` ifdefs?

juj commented:
`libc proxying`: The proxying arch exists only for some runtime functions since some runtime functions that maintain state that is now thread-local since it's in .js, but should instead be process global, will need to be proxied. I think it's just best to remove those and not keep any state in .js code since it's thread-local storage, but move all process-wide data to the Emscripten HEAP so that it's properly process global. (asm.js based VFS would fix a lot of this)

`pthreads and shared modules`: I marked the build to abort when these are attempted together. This can be worked on later.

`pthreads in the shell`: SpiderMonkey does not support web workers in a way that would allow testing this in the shell. Lars did have some hack for that, but it was too limited to be actually viable.

`musl`: That is correct. All files are taken from musl 1.0.3, and all changes should be marked with both `#ifdef __EMSCRIPTEN__` and `// XXX Emscripten`.

juj commented:
Pushed changes for all review comments so far.

kripken commented:
Ok, sounds good. LGTM.

Feel free to merge this once (1) you see all the tests passing locally, and (2) the bots are green (which seems to be the case now :) thanks!) and (3) after I have done a merge to master, as I want to merge a stable version before landing this. Now that I see the bots are green, I will focus on doing that today, hopefully no issues show up. So in summary, if you see the tests pass for you locally, and you see I've merged to master, we can then merge this.

kripken commented:
(incoming has been merged to master)

kripken commented:
Please bump the version number after you merge this (to `1.33.1`, if nothing bumps it before).

juj commented:
Great, thanks! I'm now running the full test suite for a last check, and waiting until the current tasks are finished on the bots. After those all are ok, I'll merge this in.

kripken commented:
Btw, how can the tests pass, without the fastcomp parts being merged yet?

juj commented:
Oh, I was assuming that I'd merge the fastcomp parts at the same time, they are needed as a prerequisite. Can you check that pull request and give a green light for that as well before I proceed with the merge?

kripken commented:
I must have missed an email about updates there. I'll review it now.

kripken commented:
Reviewed. Looks like we need to hold off on the merge here, there are some issues left to address on the other pull.

However, if you want to merge this with tests disabled, and enable them when the other pull merges, that sounds ok to me. It would avoid you needing to keep this mergeable.

juj commented:
The other pull request should be now in better shape. Rebased this on top of the latest incoming and reran the tests on my Windows box, which look ok.
